src/MessageSystem.hpp

```diff
@@ -0,0 +1,176 @@
+/*
+ *  Copyright (C) 2018 SPMod Development Team
+ *
+ *  This file is part of SPMod.
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include "spmod.hpp"
+
+class Message : public IMessage
+{
+public:
+    struct MessageParam
+    {
+        MsgParamType type;
+        std::variant<int, float, std::string> data;
+    };
+
+    Message() = default;
+    ~Message() = default;
+
+    // IMessage
+    std::size_t getParams() const override;
+    MsgParamType getParamType(std::size_t index) const override;
+
+    int getParamInt(std::size_t index) const override;
+    float getParamFloat(std::size_t index) const override;
+    const char *getParamString(std::size_t index) const override;
+
+    void setParamInt(std::size_t index, int value) override;
+    void setParamFloat(std::size_t index, float value) override;
+    void setParamString(std::size_t index, const char *string) override;
+
+    int getDest() const override;
+    int getType() const override;
+    const float *getOrigin() const override;
+    IEdict *getEdict() const override;
+
+    // Message
+    void init(int dest, int type, const float *origin, edict_t *edict);
+
+    void exec() const;
+
+    void clearParams();
+
+    void addParam(MsgParamType type, std::variant<int, float, std::string> &&data);
+
+    void setParam(std::size_t index, std::variant<int, float, std::string> &&data);
+
+    template<typename T>
+    const T &getParam(std::size_t index) const
```

Comment:
Do we really want a reference here?

Reply:
I got problem with strings data. Without ref in sourcepawn test plugin all string params show trash.